File: mayonez-base/src/main/java/mayonez/graphics/sprites/Sprite.java
package mayonez.graphics.sprites;

import mayonez.*;
import mayonez.graphics.*;
import mayonez.graphics.textures.*;
import mayonez.util.*;

/**
 * A visual representation of a GameObject.
 *
 * @author SlavSquatSuperstar
 */
// TODO make color and texture changeable
public abstract class Sprite extends Component {

    protected final Color color;
    protected Transform spriteXf;

    public Sprite(Color color) {
        this.color = (color != null) ? color : Colors.WHITE;
    }

    // Getters and Setters

    /**
     * Returns the color this sprite holds.
     *
     * @return the color, or white if drawing a texture
     */
    public final Color getColor() {
        return this.color;
    }

    /**
     * Get the width of this sprite's stored texture in pixels.
     *
     * @return the image width, or 0 if drawing a color
     */
    public abstract int getImageWidth();

    /**
     * Get the height of this sprite's stored texture in pixels.
     *
     * @return the image height, or 0 if drawing a color
     */
    public abstract int getImageHeight();

    /**
     * Get the pixel's RBG color on this sprite's stored texture at the specific coordinates.
     *
     * @param x the pixel x-coordinate
     * @param y the pixel y-coordinate
     * @return the color, or white if drawing a texture
     */
    public abstract Color getPixelColor(int x, int y);

    /**
     * Set the pixel's RBG color on this sprite's stored texture at the specific coordinates.
     *
     * @param x     the pixel x-coordinate
     * @param y     the pixel y-coordinate
     * @param color the color to set
     */
    public abstract void setPixelColor(int x, int y, Color color);

    /**
     * The sprite's transform in the parent object's local space.
     *
     * @return the sprite transform
     */
    public final Transform getSpriteTransform() {
        return spriteXf;
    }

    /**
     * Set additional position, rotation, and size modifiers for the sprite.
     *
     * @param spriteXf the transform
     */
    public final Sprite setSpriteTransform(Transform spriteXf) {
        this.spriteXf = spriteXf;
        return this;
    }

    /**
     * Returns the texture this sprite holds.
     *
     * @return the texture, or null if drawing a color
     */
    public abstract Texture getTexture();

    // Copy Methods

    /**
     * Returns a new sprite with the same image but not attached to any {@link mayonez.GameObject}.
     *
     * @return a copy of this image
     */
    public abstract Sprite copy();

}

File: mayonez-base/src/main/kotlin/mayonez/graphics/sprites/Sprite.kt
package mayonez.graphics.sprites

import mayonez.*
import mayonez.graphics.renderer.*
import mayonez.graphics.textures.*
import mayonez.util.*

/**
 * A visual representation of a [mayonez.GameObject]. To instantiate a
 * sprite, use [Sprites.createSprite] or [Sprites.createSprite].
 *
 * @author SlavSquatSuperstar
 */
sealed class Sprite : Component(), Renderable {

    companion object {
        @JvmStatic
        protected val DEFAULT_COLOR: MColor = Colors.WHITE
    }

    // Sprite Properties

    private var spriteXf: Transform? = null

    /**
     * Get the width of this sprite's stored texture in pixels.
     *
     * @return the image width, or 0 if drawing a color
     */
    protected abstract val imageWidth: Int

    /**
     * Get the height of this sprite's stored texture in pixels.
     *
     * @return the image height, or 0 if drawing a color
     */
    protected abstract val imageHeight: Int

    // Getters and Setters

    /**
     * Get the color of this sprite, white by default if drawing a texture.
     *
     * @return the color
     */
    abstract fun getColor(): MColor

    /**
     * Set the color of this sprite, or recolors the current texture.
     *
     * @param color the color
     */
    abstract fun setColor(color: MColor?)

    /**
     * Get the sprite's transform in the parent object's local space.
     *
     * @return the sprite transform
     */
    fun getSpriteTransform(): Transform? {
        return spriteXf
    }

    /**
     * Set additional position, rotation, and size modifiers for the sprite.
     *
     * @param spriteXf the transform
     */
    fun setSpriteTransform(spriteXf: Transform?): Sprite {
        this.spriteXf = spriteXf
        return this
    }

    /**
     * Returns the texture this sprite draws.
     *
     * @return the texture, or null if drawing a color
     */
    abstract fun getTexture(): Texture?

    /**
     * Sets the texture this sprite draws.
     *
     * @param texture the new texture
     */
    abstract fun setTexture(texture: Texture?)

    // Renderable Methods

    final override fun getZIndex(): Int {
        return gameObject.zIndex
    }

    // Copy Methods

    /**
     * Returns a new sprite with the same image but not attached to any
     * [mayonez.GameObject].
     *
     * @return a copy of this image
     */
    abstract fun copy(): Sprite?

}